### DISCUSSION TITLE
Unable to install Telnet to Ingress Controller pod

### DISCUSSION THREAD

**Discussion by Smith334** [nginx/kubernetes-ingress#4565] [Q&A] [answered]
_Background: I am using AWS and Kubernetes. I have the ingress controller running in a pod and it is proxying requests to a web service running in another pod._

What is the password of the root user account on the ingress controller pod? I am trying to install telnet so I can check connectivity, but I am unable to install it to the pod of the ingress controller.

I have tried "apt install telnet" but it returns 
```
E: Could not open lock file /var/lib/dpkg/lock-frontend - open (13: Permission denied)
E: Unable to acquire the dpkg frontend lock (/var/lib/dpkg/lock-frontend), are you root?
```

and trying "sudo apt install telnet" returns
`bash: sudo: command not found`

That's why I thought of trying to change over to the root account using "su" but I don't know the password. 
Unless someone knows of another way I can install telnet easily.
Telnet is only there for testing, so I don't want to do anything that will take too long.

**Comment by vepatel** [comment]
Hi @Smith334 pods are ephemeral so you shouldn't be installing anything inside the pod, I'd recommend prepackaging your docker image with required utilities.
```
############################################# Base image for Debian #############################################
FROM nginx:1.25.2 AS debian

RUN --mount=type=bind,from=opentracing-lib,target=/tmp/ot/ \
	apt-get update \
	&& apt-get install --no-install-recommends --no-install-suggests -y libcap2-bin \
	&& apt-get install telnet -y \
```
then
```
~/nginx/kubernetes-ingress/deployments/helm-chart on tags/v3.3.1 ● λ keti test-release-nginx-ingress-controller-84d57f79b4-dnqpq -- bash
nginx@test-release-nginx-ingress-controller-84d57f79b4-dnqpq:/$ telnet
telnet> 
```
hope this helps.

**Reply by Smith334** [reply]
Hi @vepatel . I get what you are saying that it is ephemeral, and I am sure what you said will work as I am doing something similar for another custom image that is missing packages. but I was hoping I could just install telnet for a quick single-time use, without needing to prepackage the docker image with it.
I have a complex pipeline set-up that does many things. So I would need to jerryrig the pipeline to allow for something like this ☹. 

All I am trying to do is figure out why the Nginx Ingress Controller is not proxying requests to the upstream application sometimes. It complains about a connection timeout to the upstream. The issue usually happens when I make a lot of requests very quickly through the ingress controller, and I want to know if the controller's pod can still connect to the ip and port of the upstream pod or not.

**Reply by brianehlert** [reply]
Health checks can help with that.
The N+ version of the project includes a local N+ dashboard that can show what is going on with the upstream group ( see demo.nginx.com to see this in action ).
You should be able to exec into the container ssh style.  And your K8s admins would have full control over this and no ports need to be opened.
There is also an opentracing module that is present in the image https://docs.nginx.com/nginx-ingress-controller/third-party-modules/opentracing/

I am guessing NGINX is determining that particular backend application pods are not responding and thus behaving as it always does.  Steering traffic away from the slow and not responding pods to the faster responding pods.  This all works great assuming that there are enough healthy upstream service pods to respond to all the requests.

Passive health checks are always on by default and can also be disabled.

**Reply by Smith334** [reply]
@brianehlert thank you for the reply. but right now I'm not installing Nginx Plus, I am only using the community version for now. I think it will reply on an ephemeral container like @lucacome mentioned.

**Comment by lucacome** [comment, marked as answer]
@Smith334 depending on the k8s version you're using, you might able to use `kubectl debug` with something like
```
kubectl debug -it <nginx-ingress-controller-pod-name> --image=alpine --target nginx-ingress
```

You can read more about it here: https://kubernetes.io/docs/tasks/debug/debug-application/debug-running-pod/#ephemeral-container

**Reply by Smith334** [reply]
@lucacome Something like this is what I was looking for. I'm sure it will come in handy in a few other places as well.

**Comment by jjngx** [comment]
Answered 2 years ago, closing.
